reps/2023-03-20-air-storage-path.md

```diff
@@ -0,0 +1,151 @@
+
+
+```python
+import ray
+from ray import air, tune
+
+results = tune.Tuner(train_fn, run_config=air.RunConfig(storage_path="s3://foo/bar")).fit()
+assert results.best_checkpoint().path.startswith("s3://foo/bar")
+```
+
+### General Motivation
+
+Users care about where their data is ultimately stored. They usually don't care about
+how it got there.
+
+However, currently they specify a local storage directory, from which data is periodically "synced"
+to a remote storage location. This flow (collect locally, sync to storage) is more of an implementation
+detail.
+
+Instead, users should just specify where their data ends up - with `storage_path`.
+
+In the long run, this concept helps us simplify the user journey and removes confusion
+in the configuration.
+
+Risk: Since we will continue to use locally cached data for the implementation,
+hiding the location in e.g. an environment variable can make failures more opaque
+to users.
+
+
+### Should this change be within `ray` or outside?
+
+This is an API change to Ray AIR (specifically, `air.RunConfig`, `tune.SyncConfig`, and `tune.run()`)
+and therefore has to be within Ray.
+
+
+## Stewardship
+
+### Required Reviewers
+
+@richardliaw, @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@richardliaw
+
+## Design and Architecture
+
+- We introduce a new `storage_path` argument to `air.RunConfig` and `tune.run` as the main configuration entrypoints
+- If the `storage_path` is set to a remote URI, the `local_path` is read from an environment variable `RAY_AIR_CACHE_DIR`
+- Backwards compatibility: If a `local_dir` is passed, we set this environment variable
+- In downstream components (`TrialRunner`, `Experiment`, `Trial`), we introduce respective arguments: `storage_path` and `experiment_path`
+
+We retain full backwards compatibility with the current API.
+
+## Usage Example
+
+### Basic usage
+
+```python
+import ray
+from ray import air, tune
+
+results = tune.Tuner(train_fn, run_config=air.RunConfig(storage_path="s3://foo/bar")).fit()
+assert results.best_checkpoint().path.startswith("s3://foo/bar")
+```
+
+### Usage when Ray storage is configured
+
+```python
+import ray
+from ray import air, tune
+
+# STORAGE_PATH is configured to s3://foo/bar
+
+results = tune.Tuner(train_fn).fit()
+assert results.best_checkpoint().path.startswith("s3://foo/bar")
+```
+
+### Accessing storage paths
+
+```python
+import ray
+from ray import tune
+
+results = tune.Tuner(train_fn).fit()
+
+assert results.path.startswith("s3://bucket/dir/experiment_dir")
+
+result = results.get_best_result()
+assert result.path.startswith("s3://bucket/dir/experiment_dir/trial0")
+
+checkpoint = result.checkpoint
+assert checkpoint.path.startswith("s3://bucket/dir/experiment_dir/trial0/checkpoint_0010")
```

Comment:
Can you also add examples of what happens when a local storage path is configured? Would things still magically end up on the head node, or would that not be supported?